### PR TITLE
Project Auth - Anna Hellqvist, Evelina Berg, Mats Johansen

### PR DESCRIPTION
**Pull request by AnnaHellqvist**
Link to my deployed project:
https://auth-project-api.herokuapp.com/

My reflections on how this week's project turned out:
It has been a difficult week, connecting frontend and backend. It was good to get this experience before the final project. 
It would have been good to have some information about security and authentication earlier in the sprint. It is a lot of information to understand in a week and it would have been nice to be more prepared before this project.

### CODE REVIEW

File: backend/server.js [hunk 1]
```diff
@@ -2,27 +2,129 @@ import express from 'express'
...
+      throw 'Access denied'
+    };
+    const secretMessage = `This is a secret message for ${req.user.name}`;
+    res.status(200).json({ secretMessage })
+  } catch (err) {
+    res.status(403).json({ error: 'Access Denied' });
+  }
+});
+
+
```
Comment (Gabbi-89) [inline backend/server.js:127]: Nice backend code! Easy to follow and good comments! You have used error messages and you have all needed endpoints. ⭐

File: frontend/src/App.js [hunk 2]
```diff
@@ -1,9 +1,38 @@
...
+const store = configureStore({ reducer });
 
 export const App = () => {
   return (
-    <div>
-      Find me in src/app.js!
-    </div>
+    <Provider store={store}>
+      <BrowserRouter>
+        <Switch>
```
Comment (Gabbi-89) [inline frontend/src/App.js:22]: Cool that you used the Navigation Routes and that you use reducers. 😊

File: frontend/src/lib/Container.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import styled from 'styled-components';
+
+export const Container = styled.div`
```
Comment (Gabbi-89) [inline frontend/src/lib/Container.js:3]: Nice that you have made a separate library for the styling and that you have them in different files. 💯 
Please check your indentations, you miss them for some of the styling components.

File: frontend/src/lib/InputField.js [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+`;
+
+
+const InputField = ({title, htmlFor, id, type, onChange, value}) => {
+  return (
+    <InputContainer>
+      <Label htmlFor={htmlFor}>{title}</Label>
+      <Input value={value} id={id} type={type} onChange={event => onChange(event.target.value)}></Input>
+    </InputContainer>
+  )
```
Comment (Gabbi-89) [inline frontend/src/lib/InputField.js:30]: Nice that you use props here and that you try to use it :)  Maybe it doesn't save you that much code but nice that you got it to work!

File: frontend/src/lib/Button.js [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  font-weight: bold;
+  box-shadow: 2px 2px 5px grey;
+  margin: 50px;
+
+  &:active {
+    transform: translateY(3px)
+  }
+`;
+
+export const Button = ({title, onClickFunc}) => {
```
Comment (Gabbi-89) [inline frontend/src/lib/Button.js:25]: Cool 🤩

**Review by Gabbi-89 — APPROVED** [review]
Nice project ⭐ You have fulfilled all requirements with good messages and good commented code! We really love your styling 🤩 and you did a great work having the styling in its own files. 

Good work to have separated the functions in different components which makes the code cleaner. We do also like that you used the Navigation Router! 


// Sara Stjernberg @stjernberg & Gabriella Bolin

File: backend/server.js (hunk 1 above)
Comment (Gabbi-89) [reply backend/server.js:127]: @stjernberg

File: backend/server.js [hunk 6]
```diff
@@ -2,27 +2,129 @@ import express from 'express'
...
+  try {
+    const { name, password } = req.body; // If we didnt use the schema we could do a validation here instead
+    const user = await new User({
+      name,
+      password,
+    }).save();
+    res.status(200).json({
+      userId: user._id,
+      accessToken: user.accessToken,
+      userName: user.name,
```
Comment (KarolCislowski) [inline backend/server.js:85]: You are sending all the data needed by your frontend to log in your users on spot, but you decided to force them to fill another form? It is not a nice user experience 😭

File: frontend/src/App.js [hunk 7]
```diff
@@ -1,9 +1,38 @@
...
+          <Route exact path="/">
+            <Container>
+              <LoginForm />
+              <CreateUser>
+                {/* <UserStatus /> */}
+                </CreateUser>
+            </Container>
+          </Route>
+          <Route path="/:id/user">
```
Comment (KarolCislowski) [inline frontend/src/App.js:31]: I think the better solution is to make your user component get access to the id by param or from the reducer.  🤓

File: frontend/src/components/LoginForm.js [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+            value={password}
+            aria-label='Write your password here'
+            type='password'
+            onChange={setPassword}
+          />
+          <Button title='Sign In' onClickFunc={handleLogin} />
+        </Login>
+      </LoginContainer>
+    )
+};
```
Comment (KarolCislowski) [inline frontend/src/components/LoginForm.js:81]: You have two identical forms. you could create one with two modes (in/up) passed via prop.

File: frontend/src/components/UserPage.js [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    
+    // const loginSuccess = (loginResponse) => {
+    //     dispatch(user.actions.setStatusMessage({
+    //         statusMessage: loginResponse.statusMessage
+    //     })
+    //   );
+    // };
+    // if (!accessToken) {
+    //     return <></>;
+    // };
```
Comment (KarolCislowski) [inline frontend/src/components/UserPage.js:42]: You should clean your code from commented out pieces of code 🤓

File: frontend/src/reducer/user.js [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+            const statusMessage = action.payload;
+            store.login.statusMessage = statusMessage;
+        },
+        logout: (store, action) => {
+            store.login.userId = 0;
+            store.login.accessToken = null;
+            store.login.userName = '';
+        },
+    }
+});
```
Comment (KarolCislowski) [inline frontend/src/reducer/user.js:38]: Nice that you used the redux to store your user, but you need to think about storing some of the data in the local storage, to stop logging out while refreshing the page. 🤓

**Review by KarolCislowski — CHANGES_REQUESTED** [review]
This is a nice project 🦄 but to meet the blue level you need to make your user logged in even after refreshing the page. Tips on how to do it are at the beginning of the documentation. 

Good luck 🌈

**Review by KarolCislowski — APPROVED** [review]
After the discussion with teachers about this assignment, we decided to approve this project without implementing the local storage functionality. 

However for the future, please remember to store your users' authentication data in your browser to make your UX good.

Good luck with your Final Project! 🍀

**Pull request closed without merging**